app/controllers/my_facilities/drug_stocks_controller.rb

```diff
@@ -0,0 +1,88 @@
+class MyFacilities::DrugStocksController < AdminController
+  include Pagination
+  include MyFacilitiesFiltering
+
+  layout "my_facilities"
+
+  before_action :authorize_my_facilities
+  after_action :verify_authorization_attempted
+  before_action :set_facility, only: [:new, :create]
+  before_action :set_for_end_of_month, only: [:new, :create]
+  before_action :drug_stocks_enabled?
+
+  def index
+    @facilities = filter_facilities
+      .includes(facility_group: :protocol_drugs)
+      .where(protocol_drugs: {stock_tracked: true})
+  end
+
+  def new
+    session[:report_url_with_filters] ||= request.referer
+    drug_stock_list = DrugStock.latest_for_facility(@facility, @for_end_of_month)
+    @drug_stocks = drug_stock_list.each_with_object({}) { |drug_stock, acc|
+      acc[drug_stock.protocol_drug.id] = drug_stock
+    }
+  end
+
+  def create
+    report_url_with_filters = session[:report_url_with_filters]
+    session[:report_url_with_filters] = nil
+
+    drug_stocks = DrugStock.transaction do
```

Comment:
I think this is fine, but the combination of both starting a transaction and a per record validity check feels like a lot of work in the controller.

Do you want to try moving this persistence stuff into the model, add a function that bulk saves stuff?

Reply:
We'll also look into this with https://github.com/simpledotorg/simple-server/pull/2012.